### PR TITLE
Add support for wildcard patterns in JMX dump tables configuration

### LINKED ISSUE #1564: Add wildcard support for JMX connector connector configuration.
It would greatly simplify the jmx.dump-tables property configuration of the JMX connector property file if that property supported wildcards (or regex) in way similar to how the jmx.current schema support wildcards at query time.   

See the JMX connector doc page here: https://prestosql.io/docs/current/connector/jmx.html for a description of the query time wildcard support.

### PR DESCRIPTION
**Pull request by Praveen2112**
Fixes #1564

### CODE REVIEW

File: presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxUtils.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static com.google.common.collect.Streams.stream;
+
+public class JmxUtils
+{
+    private JmxUtils()
+    {}
```
Comment (ebyhr) [inline presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxUtils.java]: nit: Most util class use `private ClassName() {}` without newline as the ctor.

Comment (Praveen2112) [reply presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxUtils.java]: Thanks. Handled it

File: presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxMetadata.java [hunk 2]
```diff
@@ -167,17 +163,6 @@ private JmxTableHandle getJmxTableHandle(SchemaTableName tableName)
         }
     }
 
-    private String toPattern(String tableName)
```
Comment (kokosing) [inline presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxMetadata.java]: why not simply make it `public static`?

File: presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxHistoricalData.java [hunk 3]
```diff
@@ -27,24 +30,36 @@
...
     }
 
-    public JmxHistoricalData(int maxEntries, Set<String> tableNames)
+    public JmxHistoricalData(int maxEntries, Set<String> tableNames, MBeanServer mbeanServer)
     {
-        tables = tableNames.stream()
-                .map(tableName -> tableName.toLowerCase(Locale.ENGLISH))
-                .collect(toSet());
+        ImmutableSet.Builder<String> builder = ImmutableSet.builder();
+        try {
```
Comment (kokosing) [inline presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxHistoricalData.java]: shouldn't this `try` be inside the `for`? Do you have a test where this exception is thrown? It would be nice to have a test with several tables where one of them (in the middle) is incorrect?

or we could simply fail the Presto by not ignoring the exception

File: presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxHistoricalData.java [hunk 4]
```diff
@@ -27,24 +30,36 @@
...
+            for (String tableName : tableNames) {
```
Comment (kokosing) [inline presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxHistoricalData.java]: `tables = tableNames.stream()...collect(toImmutableSet())`?

File: presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxHistoricalData.java [hunk 5]
```diff
@@ -19,18 +19,19 @@
...
 
+import static java.lang.management.ManagementFactory.getPlatformMBeanServer;
 import static org.testng.Assert.assertEquals;
 
 public class TestJmxHistoricalData
 {
-    private static final String TABLE_NAME = "test";
-    private static final String NOT_EXISTING_TABLE_NAME = "not-existing-test";
+    private static final String TABLE_NAME = "java.lang:name=ps old gen,type=memorypool";
+    private static final String NOT_EXISTING_TABLE_NAME = "java.lang:name=ps scavenge,type=garbagecollector";
```
Comment (kokosing) [inline presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxHistoricalData.java]: I'd undo this changes and have separate tests for for wildcards

File: presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxMetadata.java [hunk 6]
```diff
@@ -55,7 +57,11 @@
 
     private final Node localNode = createTestingNode("host1");
     private final Set<Node> nodes = ImmutableSet.of(localNode, createTestingNode("host2"), createTestingNode("host3"));
-    private final JmxMetadata metadata = new JmxMetadata(getPlatformMBeanServer(), new JmxHistoricalData(1000, ImmutableSet.of(RUNTIME_OBJECT.toLowerCase(ENGLISH))));
+    private final JmxMetadata metadata = new JmxMetadata(getPlatformMBeanServer(), new JmxHistoricalData(1000, ImmutableSet.of(RUNTIME_OBJECT.toLowerCase(ENGLISH)), getPlatformMBeanServer()));
+
+    public TestJmxMetadata()
+            throws MalformedObjectNameException
+    {}
```
Comment (kokosing) [inline presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxMetadata.java]: please remove constructor. What is the point of this?

File: presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxHistoricalData.java [hunk 7]
```diff
@@ -49,8 +53,9 @@ public void testAddingRows()
 
     @Test
     public void testCaseInsensitive()
+            throws MalformedObjectNameException
```
Comment (kokosing) [inline presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxHistoricalData.java]:
```suggestion
            throws Exception
```

File: presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxHistoricalData.java [hunk 8]
```diff
@@ -64,4 +69,18 @@ public void testCaseInsensitive()
         assertEquals(jmxHistoricalData.getRows(TABLE_NAME.toUpperCase(), columns), ImmutableList.of(
                 ImmutableList.<Object>of(42), ImmutableList.<Object>of(44)));
     }
+
+    @Test
+    public void testWildCardPatterns()
+            throws MalformedObjectNameException
```
Comment (kokosing) [inline presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxHistoricalData.java]:
```suggestion
            throws Exception
```

File: presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxHistoricalData.java (hunk 5 above)
Comment (kokosing) [reply presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxHistoricalData.java]: please restore the value for `NOT_EXISTING_TABLE_NAME`

File: presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxHistoricalData.java [hunk 9]
```diff
@@ -27,24 +30,35 @@
...
 
 public class JmxHistoricalData
 {
     private final Set<String> tables;
     private final Map<String, EvictingQueue<List<Object>>> tableData = new HashMap<>();
 
     @Inject
-    public JmxHistoricalData(JmxConnectorConfig jmxConfig)
+    public JmxHistoricalData(JmxConnectorConfig jmxConfig, MBeanServer mbeanServer)
+            throws MalformedObjectNameException
```
Comment (kokosing) [inline presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxHistoricalData.java]: I would prefer to throw PrestoException here. This exception will be burried with guice exceptions so having `PrestoException` could be a bit more user friendly.

Can you please post what kind of exception user will get during Presto initialization in case of wrong table names?

File: presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxHistoricalData.java [hunk 10]
```diff
@@ -64,4 +69,18 @@ public void testCaseInsensitive()
...
+    {
+        JmxHistoricalData jmxHistoricalData = new JmxHistoricalData(MAX_ENTRIES, ImmutableSet.of("java.lang:name=ps*"), getPlatformMBeanServer());
+
+        assertEquals(jmxHistoricalData.getTables(), ImmutableSet.of(
```
Comment (kokosing) [inline presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxHistoricalData.java]: can you also add a test where wildcard is used to query historical JMX tables?

File: presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxHistoricalData.java (hunk 9 above)
Comment (Praveen2112) [reply presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxHistoricalData.java]: This would be the exception thrown in case of wrong table names `javax.management.MalformedObjectNameException: Key properties cannot be empty`

File: presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxMetadata.java [hunk 11]
```diff
@@ -162,20 +165,24 @@ private JmxTableHandle getJmxTableHandle(SchemaTableName tableName)
...
+        try {
+            if (!tableName.contains("*")) {
+                return Pattern.quote(new ObjectName(tableName).getCanonicalName());
+            }
+            return Streams.stream(Splitter.on('*').split(tableName))
+                    .map(Pattern::quote)
+                    .collect(Collectors.joining(".*"));
+        }
+        catch (MalformedObjectNameException exception) {
+            throw new PrestoException(JMX_INVALID_TABLE_NAME, format("Not a valid ObjectName. %s", exception.getMessage()));
```
Comment (kokosing) [inline presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxMetadata.java]:
```
            throw new PrestoException(JMX_INVALID_TABLE_NAME, "Not valid object name: " + tableName);
```
?

File: presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxMetadata.java [hunk 12]
```diff
@@ -49,11 +49,12 @@
 public class TestJmxMetadata
 {
     private static final String RUNTIME_OBJECT = "java.lang:type=Runtime";
+    private static final String PATTERN = "java.lang:*";
```
Comment (kokosing) [inline presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxMetadata.java]: inline

Comment (kokosing) [reply presto-jmx/src/test/java/io/prestosql/plugin/jmx/TestJmxMetadata.java]: can you query add a test that query `jmx.history."java.lang:*"`?

File: presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxMetadata.java [hunk 13]
```diff
@@ -162,20 +167,25 @@ private JmxTableHandle getJmxTableHandle(SchemaTableName tableName)
...
+            if (!tableName.contains("*")) {
+                return Pattern.quote(new ObjectName(tableName).getCanonicalName());
+            }
+            return Streams.stream(Splitter.on('*').split(tableName))
+                    .map(Pattern::quote)
+                    .collect(Collectors.joining(".*"));
+        }
+        catch (MalformedObjectNameException exception) {
+            LOGGER.debug(exception, "Invalid ObjectName");
+            throw new PrestoException(JMX_INVALID_TABLE_NAME, format("Not a valid ObjectName %s", tableName));
```
Comment (kokosing) [inline presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxMetadata.java]: `format` is not needed here

File: presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxHistoricalData.java [hunk 14]
```diff
@@ -27,24 +29,33 @@
...
-        tables = tableNames.stream()
-                .map(tableName -> tableName.toLowerCase(Locale.ENGLISH))
-                .collect(toSet());
+        ImmutableSet.Builder<String> builder = ImmutableSet.builder();
+        for (String tableName : tableNames) {
+            String objectNamePattern = toPattern(tableName.toLowerCase(ENGLISH));
+            mbeanServer.queryNames(WILDCARD, null).stream()
+                    .map(objectName -> objectName.getCanonicalName().toLowerCase(ENGLISH))
+                    .filter(name -> name.matches(objectNamePattern))
+                    .forEach(builder::add);
```
Comment (kokosing) [inline presto-jmx/src/main/java/io/prestosql/plugin/jmx/JmxHistoricalData.java]: It looks like no exception is throw here, can you restore Java streams usage?
